File: src/account.ml
type result = Ok of int | Err of string

let string_of_result = function
  | Ok  v -> "Ok(" ^ (string_of_int v) ^ ")"
  | Err v -> "Err(" ^ v ^ ")"

let print_result r = print_endline (r |> string_of_result)

let make_withdraw balance =
  fun amount ->
    if !balance >= amount
    then (balance := !balance - amount;
          Ok !balance)
    else Err "Insufficient funds"

let () =
  let w1 = make_withdraw (ref 100)
  and w2 = make_withdraw (ref 100)
  in
  print_result (w1 10);
  print_result (w1 20);
  print_result (w1 80);
  
  print_result (w2 40)

File: src/account.rs
fn main() {
    type Account = Box<FnMut(i32) -> Result<i32, &'static str>>;

    fn make_withdraw(mut balance: i32) -> Account {
        Box::new(move |amount: i32| {
            if balance >= amount {
                balance -= amount;
                Ok(balance)
            } else {
                Err("Insufficient funds")
            }
        })
    }

    let mut w1 = make_withdraw(100);
    let mut w2 = make_withdraw(100);

    println!("{:?}", w1(10));
    println!("{:?}", w1(20));
    println!("{:?}", w1(80));

    println!("{:?}", w2(40));
}
